### PR TITLE
Save FFmpeg movies with audio

### PR DESCRIPTION
**Pull request by ggarra13**
This PR heavily modifies the FFmpegWrite.cpp to allow saving audio (or even audio only without video) with a new writeAudio() function in the Plugin API.  It also adds more extensions to those supported by our FFmpeg reader.

Audio Codec selection can be done through an ioOption.  It supports resampling of the audio as well as dealing with planar/interleaved audio and handles timecode as metadata if provided.

The code has been tested in mrv2 v0.8.3 with BART_* (44100 Hz. Mono), Dinky (48000 Hz. Stereo), elemental trailer ( 44100 Hz. Stereo with timecode), WonderWoman (44100 Hz. Stereo), dolby-spheres (96000 Hz. 7:1 8 Channels), AudioToneStereo.wav (44100 Hz Stereo no video, but currently, you still need to define a video container like .mov even if only audio gets saved).

Note: in order to handle the new FFmpeg API on Windows, it is necessary to compile FFmpegWriter.cpp with C++20, which is handled in the CMakeLists.txt file.

### CODE REVIEW

File: lib/tlIO/CMakeLists.txt [hunk 1]
```diff
@@ -67,11 +71,23 @@ if(TLRENDER_FFMPEG)
     list(APPEND SOURCE FFmpeg.cpp FFmpegRead.cpp FFmpegReadAudio.cpp
         FFmpegReadVideo.cpp FFmpegWrite.cpp)
     list(APPEND LIBRARIES_PRIVATE FFmpeg)
+    if(WIN32)
+	set_source_files_properties(FFmpegWrite.cpp
```
Comment (darbyjohnston) [inline lib/tlIO/CMakeLists.txt:75]: Why is the C++20 necessary?

File: lib/tlIO/CMakeLists.txt [hunk 2]
```diff
@@ -67,11 +71,23 @@ if(TLRENDER_FFMPEG)
...
+	    PROPERTIES COMPILE_FLAGS -std:c++20)
+    endif()
 endif()
 if(TLRENDER_USD)
     list(APPEND HEADERS_PRIVATE USD.h USDPrivate.h)
     list(APPEND SOURCE USD.cpp USDRead.cpp USDRender.cpp)
     list(APPEND LIBRARIES_PRIVATE tf usd usdGeom usdImagingGL usdAppUtils hd)
+    if(NOT WIN32)
+	set_source_files_properties(USD.cpp
```
Comment (darbyjohnston) [inline lib/tlIO/CMakeLists.txt:84]: Please no not include unrelated changes in the PR.

File: lib/tlIO/FFmpeg.cpp [hunk 3]
```diff
@@ -177,16 +190,33 @@ namespace tl
             IPlugin::_init(
                 "FFmpeg",
                 {
-                    { ".mov", io::FileType::Movie },
+                    // Video Formats
+                    { ".avi", io::FileType::Movie },
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpeg.cpp:194]: This should be a separate PR. Also as I have mentioned before, I do not want to add support for older/non-VFX related formats like webm and flv.

File: lib/tlIO/FFmpegWrite.cpp [hunk 4]
```diff
@@ -2,32 +2,150 @@
...
+                }
+            }
+            
+            //! Check that a given sample format is supported by the encoder
+            bool checkSampleFormat(
+                const AVCodec* codec, enum AVSampleFormat sample_fmt)
+            {
+                const enum AVSampleFormat *p = codec->sample_fmts;
+
+                while (*p != AV_SAMPLE_FMT_NONE) {
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpegWrite.cpp:55]: Please format the code consistently like the other code in the file.

File: lib/tlIO/FFmpegWrite.cpp [hunk 5]
```diff
@@ -350,40 +925,204 @@ namespace tl
...
+                    audio = planarDeinterleave(audioResampled);
+                else
+                    audio = audioResampled;
+                
+                uint8_t* data = audio->getData();
+                
+                // Allocate flatData pointers
+                const size_t channels = audio->getChannelCount();
+                const size_t stride = audio->getByteCount() / channels;
+                uint8_t** flatData = new uint8_t*[channels];
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpegWrite.cpp:986]: How is this memory cleaned up?

File: lib/tlIO/CMakeLists.txt (hunk 1 above)
Comment (ggarra13) [reply lib/tlIO/CMakeLists.txt:72]: C++ is needed as MSVC (2019 at least), cannot copy or take an address of these macros:

AVChannelLayout channel_layout = AV_CHANNEL_LAYOUT_STEREO;
return av_channel_layout_copy(dst, &channel_layout);

Linux and macOS accept that already.

File: lib/tlIO/CMakeLists.txt (hunk 2 above)
Comment (ggarra13) [reply lib/tlIO/CMakeLists.txt]: Removed.

File: lib/tlIO/FFmpeg.cpp (hunk 3 above)
Comment (ggarra13) [reply lib/tlIO/FFmpeg.cpp]: I reverted to your list of formats only.

File: lib/tlIO/FFmpegWrite.cpp (hunk 4 above)
Comment (ggarra13) [reply lib/tlIO/FFmpegWrite.cpp:55]: Hopefully fixed.

File: lib/tlIO/FFmpegWrite.cpp (hunk 5 above)
Comment (ggarra13) [reply lib/tlIO/FFmpegWrite.cpp]: OOPS!!! I changed it to use a std::vector in private data.

File: lib/tlIO/CMakeLists.txt (hunk 1 above)
Comment (darbyjohnston) [reply lib/tlIO/CMakeLists.txt:72]: I think C++20 as a requirement is too early, it's not on the VFX reference platform yet. Hopefully there is another way to do this without requiring C++20.

Comment (ggarra13) [reply lib/tlIO/CMakeLists.txt]: Yes.  I found out that the function av_channel_layout_default(dst, channelCount); handles MONO and STEREO just fine, so that my checks were unneded.  Removed the C++20 dependency.

File: lib/tlIO/FFmpegWrite.cpp (hunk 4 above)
Comment (darbyjohnston) [reply lib/tlIO/FFmpegWrite.cpp:55]: For example:
* Always initialize variables when they are declared, and declare just one variable per line:
```
const AVChannelLayout* p = nullptr;
const AVChannelLayout* best_ch_layout = nullptr;
```
* Try and have a single return statement in functions.
* Brackets are on their own lines:
```
if (...)
{
    ...
}
```

File: lib/tlIO/FFmpegWrite.cpp [hunk 6]
```diff
@@ -2,32 +2,166 @@
...
+                return out;
+            }
+            
+            //! Return an equal or higher supported samplerate
+            int selectSampleRate(const AVCodec* codec, const int sampleRate)
+            {
+                int out = 0;
+                if (!codec->supported_samplerates)
+                {
+                    out = 44100;
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpegWrite.cpp:115]: It would be nice to replace this and the other two occurrences with a constant:
```
const size_t sampleRateDefault = 44100;
```

File: lib/tlIO/FFmpegWrite.cpp [hunk 7]
```diff
@@ -42,198 +176,582 @@ namespace tl
...
-            //av_dump_format(p.avFormatContext, 0, p.fileName.c_str(), 1);
+                p.sampleRate =
+                    selectSampleRate(avCodec, info.audio.sampleRate);
+                if (p.sampleRate == 0)
+                    throw std::runtime_error(
+                        string::Format(
+                            "{0}: Could not select sample rate")
+                            .arg(p.fileName));
+                    
+                char buf[256];
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpegWrite.cpp:382]: It would be nice to use a std::vector instead of C style array here. If the size needs to change later it makes it easier to update:
```
std::vector<char> buf(256);
av_channel_layout_describe(&p.avAudioCodecContext->ch_layout, buf.data(), buf.size());
```

File: lib/tlIO/FFmpegWrite.cpp [hunk 8]
```diff
@@ -42,198 +176,582 @@ namespace tl
...
+                            .arg(input.sampleRate));
+                    }
+                }
 
-            r = avio_open(&p.avFormatContext->pb, p.fileName.c_str(), AVIO_FLAG_WRITE);
-            if (r < 0)
-            {
-                throw std::runtime_error(string::Format("{0}: {1}").arg(p.fileName).arg(getErrorLabel(r)));
-            }
+                p.avAudioCodecContext->bit_rate = 69000;
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpegWrite.cpp:429]: I like to try and put hard-coded numbers in an anonymous namespace at the top of the file for visibility:
```
namespace
{
    const size_t bitRateDefault = 69000;
}
```

File: lib/tlIO/FFmpegWrite.cpp [hunk 9]
```diff
@@ -42,198 +176,582 @@ namespace tl
...
+                p.avAudioCodecContext->sample_rate = p.sampleRate;
+                p.avAudioCodecContext->time_base.num = 1;
+                p.avAudioCodecContext->time_base.den = p.sampleRate;
 
-            r = avformat_write_header(p.avFormatContext, NULL);
-            if (r < 0)
-            {
-                throw std::runtime_error(string::Format("{0}: {1}").arg(p.fileName).arg(getErrorLabel(r)));
-            }
+                if ((p.avAudioCodecContext->block_align == 1 ||
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpegWrite.cpp:434]: What are these numbers?

File: lib/tlIO/FFmpegWrite.cpp [hunk 10]
```diff
@@ -42,198 +176,582 @@ namespace tl
...
+                case image::PixelType::RGB_U8:   p.avPixelFormatIn = AV_PIX_FMT_RGB24;  break;
+                case image::PixelType::RGBA_U8:  p.avPixelFormatIn = AV_PIX_FMT_RGBA;   break;
+                case image::PixelType::L_U16:    p.avPixelFormatIn = AV_PIX_FMT_GRAY16; break;
+                case image::PixelType::RGB_U16:  p.avPixelFormatIn = AV_PIX_FMT_RGB48;  break;
+                case image::PixelType::RGBA_U16: p.avPixelFormatIn = AV_PIX_FMT_RGBA64; break;
+                default:
+                    throw std::runtime_error(string::Format("{0}: Incompatible pixel type").arg(p.fileName));
+                    break;
+                }
+                /*p.swsContext = sws_getContext(
```
Comment (darbyjohnston) [inline lib/tlIO/FFmpegWrite.cpp:697]: This commented out code could be removed, I only kept it temporarily to make sure the `sws_alloc_context()` worked as a replacement.

File: lib/tlIO/FFmpegWrite.cpp (hunk 9 above)
Comment (ggarra13) [reply lib/tlIO/FFmpegWrite.cpp:434]: To be quite honest, I have no clue.  They appear in ffmpeg.c without any documentation.
